src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs

```diff
@@ -1457,6 +1494,1629 @@ class Test : System.Attribute
+            CompileAndVerify(compilation, expectedOutput:
+@"System.InvalidOperationException
+System.InvalidOperationException");
+        }
+
+        [Fact]
+        public void Catch_02()
+        {
+            var source =
+@"
+public class X
+{
+    public static void Main()
+    {
+        try
+        {
+            throw new System.InvalidOperationException();
+        }
+        catch (System.Exception e) when (Dummy(TakeOutParam(e, out var x1), x1))
+        {
+            System.Action d = () =>
+                                {
+                                    System.Console.WriteLine(x1.GetType());
+                                };
+
+            System.Console.WriteLine(x1.GetType());
+            d();
+        }
+    }
+
+    static bool Dummy(object y, object z) 
+    {
+        System.Console.WriteLine(z.GetType());
+        return true;
+    }
+
+    static bool TakeOutParam<T>(T y, out T x) 
+    {
+        x = y;
+        return true;
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithOutVarFeature());
+            CompileAndVerify(compilation, expectedOutput:
+@"System.InvalidOperationException
+System.InvalidOperationException
+System.InvalidOperationException");
+        }
+
+        [Fact]
+        public void Catch_03()
+        {
+            var source =
+@"
+public class X
+{
+    public static void Main()
+    {
+        try
+        {
+            throw new System.InvalidOperationException();
+        }
+        catch (System.Exception e) when (Dummy(TakeOutParam(e, out var x1), x1))
+        {
+            System.Action d = () =>
+                                {
+                                    e = new System.NullReferenceException();
+                                    System.Console.WriteLine(x1.GetType());
+                                };
+
+            System.Console.WriteLine(x1.GetType());
+            d();
+            System.Console.WriteLine(e.GetType());
+        }
+    }
+
+    static bool Dummy(object y, object z) 
+    {
+        System.Console.WriteLine(z.GetType());
+        return true;
+    }
+
+    static bool TakeOutParam<T>(T y, out T x) 
+    {
+        x = y;
+        return true;
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithOutVarFeature());
+            CompileAndVerify(compilation, expectedOutput:
+@"System.InvalidOperationException
+System.InvalidOperationException
+System.InvalidOperationException
+System.NullReferenceException");
+        }
+
+        [Fact]
+        public void Catch_04()
+        {
+            var source =
+@"
+public class X
+{
+    public static void Main()
+    {
+        try
+        {
+            throw new System.InvalidOperationException();
+        }
+        catch (System.Exception e) when (Dummy(TakeOutParam(e, out var x1), x1))
```

Comment:
These are good suggestions. In this particular case it is not likely to run into problems because exception variable is sharing the scope and we should be able to capture it. I added a work item into https://github.com/dotnet/roslyn/issues/11566 to make sure we have a good coverage for lambdas and await.